### PR TITLE
Derive hash functions for the generic AST

### PR DESCRIPTION
**Pull request by mjambon**
This is meant for the caching optimizations in semgrep.

### CODE REVIEW

**Comment by mjambon** [comment]
Please don't merge yet, since it will break semgrep-core.

**Review by aryx — APPROVED** [review]

**Comment by aryx** [comment]
We could merge anyway since semgrep-core uses a pinned version of pfff.

**Review by aryx — APPROVED** [review]

File: lang_GENERIC_base/AST_generic.ml [hunk 1]
```diff
@@ -754,20 +787,27 @@ and other_expr_operator =
...
+(*
+   Two statements are considered equal iff their ID ('s_id' field) is the
+   same. The 'equal_stmt' and 'hash_stmt' functions work accordingly.
+*)
 (*s: type [[AST_generic.stmt]] *)
 and stmt = {
-  s: stmt_kind;
-  (* this can be used to hash more efficiently stmts, or in semgrep to
-   * quickly know if a stmt is a children of another stmt.
+  s: stmt_kind [@equal fun _a _b -> true] [@hash.ignore];
```
Comment (aryx) [inline lang_GENERIC_base/AST_generic.ml]: Cant do that as i explained on slack.

**Review by aryx — CHANGES_REQUESTED** [review]

File: lang_GENERIC_base/AST_generic.ml [hunk 2]
```diff
@@ -140,22 +140,124 @@
...
+(*
+   Create a node ID that is guaranteed to be unique within an AST,
+   as long as the whole AST is created within the same process.
+
+   Such ID should not be expected to be unique across ASTs,
+   since ASTs can be loaded from an external cache, for example.
+*)
+module Node_ID :
+sig
+  type t = private int [@@deriving show, eq, hash]
```
Comment (aryx) [inline lang_GENERIC_base/AST_generic.ml]: This is not supported by the ocaml parser in pfff, and so semgrep choke in CI. Can you do without
this advanced feature?

**Review by aryx — APPROVED** [review]
Looks good, but if you could remove the private keyword so CI does not fail
(or fix lang_ocaml/parsing/parser_ml.mly)

Is there a way you could move out some of the code out of AST_generic.ml?
I try to keep AST_generic.ml as small as possible and move helpers functions
out in AST_generic_helpers.ml when I can.

File: lang_GENERIC_base/AST_generic.ml [hunk 3]
```diff
@@ -140,22 +140,124 @@
...
+let with_structural_equal equal a b =
+  match !busy_with_equal with
+  | Not_busy ->
+      busy_with_equal := Structural_equal;
+      Fun.protect
+        ~finally:(fun () -> busy_with_equal := Not_busy)
+        (fun () -> equal a b)
+  | Structural_equal
+  | Referential_equal ->
+      failwith "an equal is already in progress"
```
Comment (aryx) [inline lang_GENERIC_base/AST_generic.ml:245]: This busy trick is nice! I thought about using a global, but didn't think about how
to make sure we don't use both at the same time (e.g., in different threads or later different cores).

**Review by aryx — APPROVED** [review]
ok let's merge this!

**Comment by mjambon** [comment]
There's a problem an ubuntu repo. Is this new?

**Review by aryx — APPROVED** [review]

**Pull request merged**